source/Calamari/Kubernetes/ResourceStatus/ResourceUpdateReporter.cs

```diff
@@ -0,0 +1,96 @@
+
+namespace Calamari.Kubernetes.ResourceStatus
+{
+    public interface IResourceUpdateReporter
+    {
+        /// <summary>
+        /// Reports the difference of the originalStatuses and newStatuses to server.
+        /// </summary>
+        void ReportUpdatedResources(IDictionary<string, Resource> originalStatuses, IDictionary<string, Resource> newStatuses);
+    }
+    
+    /// <summary>
+    /// <inheritdoc />
+    /// </summary>
+    public class ResourceUpdateReporter : IResourceUpdateReporter
+    {
+        private readonly IVariables variables;
+        private readonly ILog log;
+        
+        public ResourceUpdateReporter(IVariables variables, ILog log)
+        {
+            this.variables = variables;
+            this.log = log;
+        }
+        
+        public void ReportUpdatedResources(IDictionary<string, Resource> originalStatuses, IDictionary<string, Resource> newStatuses)
+        {
+            var createdOrUpdatedResources = GetCreatedOrUpdatedResources(originalStatuses, newStatuses);
+            var removedResources = GetRemovedResources(originalStatuses, newStatuses);
+            foreach (var resource in createdOrUpdatedResources.Concat(removedResources))
+            {
+                SendServiceMessage(resource);
+            }
+        }
+        
+        private static IEnumerable<Resource> GetCreatedOrUpdatedResources(IDictionary<string, Resource> originalStatuses, IDictionary<string, Resource> newStatuses)
+        {
+            var createdOrUpdated = new List<Resource>();
+            foreach (var resource in newStatuses)
+            {
+                if (!originalStatuses.ContainsKey(resource.Key) || resource.Value.HasUpdate(originalStatuses[resource.Key]))
+                {
+                    createdOrUpdated.Add(resource.Value);
+                }
+            }
+            return createdOrUpdated;
+        }
+
+        private static IEnumerable<Resource> GetRemovedResources(IDictionary<string, Resource> originalStatuses, IDictionary<string, Resource> newStatuses)
+        {
+            var removed = new List<Resource>();
+            foreach (var resource in originalStatuses)
+            {
+                if (!newStatuses.ContainsKey(resource.Key))
+                {
+                    resource.Value.Removed = true;
+                    removed.Add(resource.Value);
+                }
+            }
+            return removed;
+        }
+        
+        private void SendServiceMessage(Resource resource)
+        {
+            var parameters = new Dictionary<string, string>
+            {
+                {"type", "k8s-status"},
+                {"actionId", variables.Get("Octopus.Action.Id")},
+                {"taskId", variables.Get(KnownVariables.ServerTask.Id)},
+                {"targetId", variables.Get("Octopus.Machine.Id")},
+                {"environmentId", variables.Get(DeploymentEnvironment.Id)},
+                {"spaceId", variables.Get("Octopus.Space.Id")},
```

Comment:
It feels like this operation shouldnt care about spaceId, environmentId etc. Thats just a result of the task its being logged against.
We should ideally be able to determine this from the context of the operation over in Octopus

Reply:
@johnsimons Based on your work on the database side, shall we remove some of these IDs if they are not used?

Reply:
I am using all of these @YihaoOct

Reply:
Given we have the info here, I don't see the reason to add more work to the server to do these lookups.
@zentron apart from sending this info across the wire, is there a technical reason to remove these?

Reply:
More an architectural reason. Its possible that the task might not be running for a project, or at some point even an environment. It also feels like a bunch of redundant information given its already connected via the server task. 
Lets catch up this morning.